File: articles/calculus-for-systems-modeling/articles/rules-of-differentiation-and-model-structure/julia/structural_derivative_audit.jl
population(t) = 100.0 * exp(0.01 * t)
population_rate(t) = 0.01 * population(t)
affluence(t) = 2.0 * exp(0.02 * t)
affluence_rate(t) = 0.02 * affluence(t)

println("rule,model_structure,t,derivative_value,component_a,component_b,warning")
for t in [0.0, 5.0, 10.0, 20.0]
    a = population_rate(t) * affluence(t)
    b = population(t) * affluence_rate(t)
    println("product_rule,impact = population * affluence,$t,$(a+b),$a,$b,")
end

File: articles/calculus-for-systems-modeling/articles/rules-of-differentiation-and-model-structure/cpp/structural_derivative_audit.cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <vector>

double population(double t){ return 100.0 * std::exp(0.01 * t); }
double population_rate(double t){ return 0.01 * population(t); }
double affluence(double t){ return 2.0 * std::exp(0.02 * t); }
double affluence_rate(double t){ return 0.02 * affluence(t); }

int main(){
  std::vector<double> ts={0.0,5.0,10.0,20.0};
  std::cout<<std::fixed<<std::setprecision(12);
  std::cout<<"rule,model_structure,t,derivative_value,component_a,component_b,warning\n";
  for(double t:ts){
    double a=population_rate(t)*affluence(t);
    double b=population(t)*affluence_rate(t);
    std::cout<<"product_rule,impact = population * affluence,"<<t<<","<<a+b<<","<<a<<","<<b<<",\n";
  }
}
